File: backend/app/models/sos.py
import uuid
from datetime import datetime, timezone
import enum
from sqlalchemy import Column, String, Float, DateTime, ForeignKey, Enum as SQLEnum, Integer, Text
from sqlalchemy.orm import relationship
from backend.app.db.session import Base


class SOSStatus(str, enum.Enum):
    CREATED = "CREATED"
    ACTIVE = "ACTIVE"
    ACKNOWLEDGED = "ACKNOWLEDGED"
    RESOLVED = "RESOLVED"
    CANCELLED = "CANCELLED"
    EXPIRED = "EXPIRED"


class SOSEvent(Base):
    __tablename__ = "sos_events"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    status = Column(SQLEnum(SOSStatus), default=SOSStatus.ACTIVE, nullable=False, index=True)
    latitude = Column(Float, nullable=False)
    longitude = Column(Float, nullable=False)
    address_approx = Column(String(255), nullable=True)
    triggered_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    resolved_at = Column(DateTime, nullable=True)
    contacts_notified_count = Column(Integer, default=0, nullable=False)
    notification_status = Column(String(100), default="SENT", nullable=False)  # SENT, FAILED, PARTIAL
    admin_notes = Column(Text, nullable=True)
    responder_id = Column(String(36), nullable=True)
    accuracy_meters = Column(Float, default=10.0, nullable=False)
    journey_id = Column(String(36), ForeignKey("journeys.id", ondelete="SET NULL"), nullable=True, index=True)
    share_token = Column(String(128), nullable=True)

    user = relationship("User", back_populates="sos_events")
    journey = relationship("Journey")

    @property
    def user_name(self) -> Optional[str]:
        return self.user.full_name if self.user else None

    @property
    def user_phone(self) -> Optional[str]:
        return self.user.phone if self.user else None

    @property
    def destination_name(self) -> Optional[str]:
        return self.journey.destination_name if self.journey else None

File: mobile/lib/models/sos.dart
class SOSEvent {
  final String id;
  final String userId;
  final String status;
  final double latitude;
  final double longitude;
  final String? addressApprox;
  final DateTime triggeredAt;
  final DateTime? resolvedAt;
  final int contactsNotifiedCount;
  final String notificationStatus;
  final String? adminNotes;

  SOSEvent({
    required this.id,
    required this.userId,
    required this.status,
    required this.latitude,
    required this.longitude,
    this.addressApprox,
    required this.triggeredAt,
    this.resolvedAt,
    required this.contactsNotifiedCount,
    required this.notificationStatus,
    this.adminNotes,
  });

  factory SOSEvent.fromJson(Map<String, dynamic> json) {
    return SOSEvent(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      status: json['status'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      addressApprox: json['address_approx'] as String?,
      triggeredAt: DateTime.parse(json['triggered_at'] as String),
      resolvedAt: json['resolved_at'] != null ? DateTime.parse(json['resolved_at'] as String) : null,
      contactsNotifiedCount: json['contacts_notified_count'] as int? ?? 0,
      notificationStatus: json['notification_status'] as String? ?? 'SENT',
      adminNotes: json['admin_notes'] as String?,
    );
  }
}
